File: dss/general/DSSObject.py
# Copyright (C) 2010 Richard Lincoln
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA, USA

from dss.common.Named import Named

from dss.common.Utilities import CheckForBlanks

class DSSObject(Named):

    def ClearPropSeqArray(self):
        self.PropSeqCount = 0
        for i in range(self.ParentClass.NumProperties):
            self.PrpSequence[i] = 0


    def __init__(self, ParClass, DSSObjType=0, DSSClassName='', classIndex=0,
            dirty=False, flag=False, propSequence=0.0, parentClass=None):
        """Initialises a new 'DSSObject' instance.
        """
        super(self).__init__(ParClass.Name)

        self.DSSObjType = 0  # PD, PC, Monitor, CondCode, etc.
        self.ParentClass = ParClass
        self.ClassIndex = 0
        self.HasBeenSaved = False
        #: General purpose Flag for each object  don't assume inited.
        self.Flag = False

        #: PD, PC, Monitor, CondCode, etc.
        self.DSSObjType = DSSObjType

        self.DSSClassName = DSSClassName

        #: Index into the class collection list.
        self.classIndex = classIndex

        self.dirty = dirty

        #: General purpose Flag for each object - don't assume inited.
        self.flag = flag

        self.propSequence = propSequence

        self.parentClass = [] if parentClass is None else parentClass

        ## Protected
        self.PropSeqCount = 0
        self.PropertyValue = []
        self.PrpSequence = []


    def DumpProperties(self, F, Complete=False):
        F.write('\n')
        F.write('New ' + self.DSSClassName + '.' + self.Name + '\n')


    def Edit(self):
        """Allow Calls to edit from object itself."""
        self.ParentClass.Active = self.ClassIndex
        return self.ParentClass.Edit()


    def GetPropertyValue(self, Index=0):
        # Use dssclass.propertyindex to get index by name
        # Default Behavior for all DSS Objects
        return self.PropertyValue[Index]


    def _Get_PropertyValue(self, Index=0):
        # This is virtual function that may call routine
        return self.GetPropertyValue(Index)

    def _Set_PropertyValue(self, Index=0, Value=""):
        self.PropertyValue[Index] = Value

        # Keep track of the order in which this property was accessed
        # for Save Command
        self.PropSeqCount += 1
        self.PrpSequence[Index] = self.PropSeqCount

    PropertyValue = property(_Get_PropertyValue, _Set_PropertyValue)


    def InitPropertyValues(self, ArrayOffset=0):
        self.PropertyValue[ArrayOffset + 1] = ''  # Like Property

        # Clear propertySequence Array  after initialization
        self.ClearPropSeqArray()


    def SaveWrite(self, F):
        # Write only properties that were explicitly set in the
        # final order they were actually set}
        iProp = self.GetNextPropertySet(0)  # Works on ActiveDSSObject
        while iProp > 0:
            pc = self.ParentClass
            F.write(' ', pc.PropertyName[pc.RevPropertyIdxMap[iProp]])
            F.write('=', CheckForBlanks( self.PropertyValue[iProp] ))
            iProp = self.GetNextPropertySet(iProp)


    def GetNextPropertySet(self, idx=0):
        """Find next larger property sequence number
        return 0 if none found"""
        Smallest = 9999999  # some big number

        if idx > 0:
            idx = self.PrpSequence[idx]
        for i in range(self.ParentClass.NumProperties):
            if self.PrpSequence[i] > idx:
                if self.PrpSequence[i] < Smallest:
                    Smallest = self.PrpSequence[i]
                    return i
        return 0


    def _Get_Name(self):
        return self.LocalName

    def _Set_Name(self, Value=""):
        """If renamed, then let someone know so hash list can be updated"""
        if len(self.LocalName) > 0:
            self.ParentClass.ElementNamesOutOfSynch = True
        self.LocalName = Value

    Name = property(_Get_Name, _Set_Name)

File: dss/src/main/java/com/ncond/dss/general/DSSObject.java
/**
 * Copyright (C) 2008-2012 Electric Power Research Institute, Inc.
 * Copyright (C) 2009-2012 Richard Lincoln
 * All rights reserved.
 */
package com.ncond.dss.general;

import java.io.OutputStream;
import java.io.PrintWriter;

import com.ncond.dss.common.DSSClass;
import com.ncond.dss.common.Util;

abstract public class DSSObject extends NamedObject {

	protected int propSeqCount;

	protected String[] propertyValues;

	protected int[] prpSequence;

	/** PD, PC, Monitor, CondCode, etc. */
	protected int objType;

	protected DSSClass parentClass;

	/** Index into the class collection list */
	protected int classIndex;

	protected boolean hasBeenSaved;

	/** General purpose flag for each object  don't assume inited */
	protected boolean flag;

	public void clearPropSeqArray() {
		propSeqCount = 0;
		for (int i = 0; i < parentClass.getNumProperties(); i++) {
			prpSequence[i] = 0;
		}
	}

	public DSSObject(DSSClass parClass) {
		super(parClass.getClassName());

		objType = 0;
		propSeqCount = 0;
		parentClass = parClass;
		propertyValues = new String[parentClass.getNumProperties()];

		// init'd to zero when allocated
		prpSequence = new int[parentClass.getNumProperties()];

		hasBeenSaved = false;
	}

	public void dumpProperties(OutputStream out, boolean complete) {
		PrintWriter pw = new PrintWriter(out);
		pw.println();
		pw.println("new " + getDSSClassName() + '.' + getName());
		pw.close();
	}

	/**
	 * Allow calls to edit from object itself.
	 */
	public int edit() {
		parentClass.setActiveElement(classIndex);
		return parentClass.edit();
	}

	/**
	 * Use DSSClass.propertyIndex() to get index by name.
	 */
	public String getPropertyValue(int index) {
		// default behavior for all DSS objects
		return propertyValues[index];
	}

	public void initPropertyValues(int arrayOffset) {
		propertyValues[arrayOffset] = "";

		// clear propertySequence array after initialization
		clearPropSeqArray();
	}

	public void saveWrite(PrintWriter f) {
		DSSClass pc;

		/* Write only properties that were explicitly set in the
		 * final order they were actually set. */
		int iProp = getNextPropertySet(0); // works on activeDSSObject

		while (iProp >= 0) {
			pc = parentClass;
			f.print(" " + pc.getPropertyName(pc.getRevPropertyIdxMap(iProp)));
			f.print("=" + Util.checkForBlanks( propertyValues[iProp] ));
			iProp = getNextPropertySet(iProp);
		}
	}

	/**
	 * Find next larger property sequence number
	 * return -1 if none found
	 */
	protected int getNextPropertySet(int idx) {
		int smallest = 9999999;  // some big number
		int result = -1;

		if (idx >= 0) idx = prpSequence[idx];

		for (int i = 0; i < parentClass.getNumProperties(); i++)
			if (prpSequence[i] > idx + 1)  // one-based
				if (prpSequence[i] < smallest) {
					smallest = prpSequence[i];
					result = i;
				}

		return result;
	}

	public void setName(String value) {
		// if renamed, then let someone know so hash list can be updated
		if (getLocalName().length() > 0)
			parentClass.setElementNamesOutOfSynch(true);

		setLocalName(value);
	}

	public void setPropertyValue(int index, String value) {
		propertyValues[index] = value;

		// keep track of the order in which this property was
		// accessed for save command
		propSeqCount += 1;
		prpSequence[index] = propSeqCount;
	}

	public String getName() {
		return getLocalName();
	}

	public String[] getPropertyValue() {
		return propertyValues;
	}

	public int getObjType() {
		return objType;
	}

	public int getClassIndex() {
		return classIndex;
	}

	public DSSClass getParentClass() {
		return parentClass;
	}

	public boolean isHasBeenSaved() {
		return hasBeenSaved;
	}

	public void setHasBeenSaved(boolean hasBeenSaved) {
		this.hasBeenSaved = hasBeenSaved;
	}

	public boolean isFlag() {
		return flag;
	}

	public void setFlag(boolean flag) {
		this.flag = flag;
	}

	public int[] getPrpSequence() {
		return prpSequence;
	}

	public void setClassIndex(int classIndex) {
		this.classIndex = classIndex;
	}

}
